File: src/lib/apiCrypto.ts
/**
 * Decrypts AES-256-GCM envelopes returned by the backend ({ iv, data }, both
 * base64) and provides a fetch wrapper for admin API calls.
 *
 * The key is derived (SHA-256) from a secret shared with the backend and the
 * Flutter app. Because this runs in the browser, the secret ships inside the
 * JS bundle and isn't truly hidden from someone reading the code — but the
 * GCM auth tag still guarantees tamper-evidence: any modified ciphertext
 * fails to decrypt instead of silently applying a changed value.
 */
const SHARED_SECRET = 'AshutoshLaw#Secure2026$VPS-Encryption-Key';

let cachedKey: CryptoKey | null = null;

async function getKey(): Promise<CryptoKey> {
  if (cachedKey) return cachedKey;
  const secretBytes = new TextEncoder().encode(SHARED_SECRET);
  const hash = await crypto.subtle.digest('SHA-256', secretBytes);
  cachedKey = await crypto.subtle.importKey('raw', hash, 'AES-GCM', false, ['decrypt']);
  return cachedKey;
}

function base64ToBytes(b64: string): Uint8Array {
  const bin = atob(b64);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

export async function decryptEnvelope<T = unknown>(envelope: { iv: string; data: string }): Promise<T> {
  const key = await getKey();
  const iv = base64ToBytes(envelope.iv);
  const ciphertext = base64ToBytes(envelope.data);
  const plaintextBuf = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: iv as BufferSource },
    key,
    ciphertext as BufferSource
  );
  const text = new TextDecoder().decode(plaintextBuf);
  return JSON.parse(text) as T;
}

/**
 * Fetches JSON from an admin API endpoint and decrypts it.
 * Throws on network failure or a failed decrypt (tampered/corrupt payload) —
 * callers should treat that as "couldn't reach the server", never as
 * "user is unauthorized".
 */
export async function fetchEncryptedJson<T = unknown>(input: RequestInfo | URL, init?: RequestInit): Promise<T> {
  const res = await fetch(input, init);
  if (!res.ok) {
    throw new Error(`Request failed with status ${res.status}`);
  }
  const envelope = await res.json();
  return decryptEnvelope<T>(envelope);
}

File: admin_mobile_app/lib/services/api_crypto.dart
import 'dart:convert';
import 'package:cryptography/cryptography.dart';
import 'package:http/http.dart' as http;

/// Decrypts the AES-256-GCM envelopes ({ iv, data }, both base64) returned
/// by the backend for admin endpoints. The key is derived (SHA-256) from a
/// secret shared with the backend and the web admin console — see
/// backend-api/src/crypto.rs and src/lib/apiCrypto.ts for the matching
/// implementations.
class ApiCrypto {
  static const String _sharedSecret = 'AshutoshLaw#Secure2026\$VPS-Encryption-Key';
  static SecretKey? _cachedKey;

  static Future<SecretKey> _getKey() async {
    if (_cachedKey != null) return _cachedKey!;
    final hash = await Sha256().hash(utf8.encode(_sharedSecret));
    _cachedKey = SecretKey(hash.bytes);
    return _cachedKey!;
  }

  static Future<dynamic> decryptEnvelope(Map<String, dynamic> envelope) async {
    final algorithm = AesGcm.with256bits();
    final key = await _getKey();

    final ivBytes = base64.decode(envelope['iv'] as String);
    final full = base64.decode(envelope['data'] as String);
    // AES-GCM tag is the last 16 bytes, appended by the encrypting side.
    final cipherBytes = full.sublist(0, full.length - 16);
    final macBytes = full.sublist(full.length - 16);

    final secretBox = SecretBox(cipherBytes, nonce: ivBytes, mac: Mac(macBytes));
    final clearBytes = await algorithm.decrypt(secretBox, secretKey: key);
    return json.decode(utf8.decode(clearBytes));
  }
}

class ApiCryptoException implements Exception {
  final String message;
  ApiCryptoException(this.message);
  @override
  String toString() => message;
}

/// Thin http helper: GET/POST/DELETE an admin endpoint and decrypt the
/// response body. Any failure here (network, non-200, bad decrypt) throws —
/// callers must treat that as "couldn't reach the server", not "unauthorized".
class ApiClient {
  static Future<dynamic> get(String url) async {
    final res = await http.get(Uri.parse(url));
    if (res.statusCode != 200) {
      throw ApiCryptoException('Server error (${res.statusCode})');
    }
    return ApiCrypto.decryptEnvelope(json.decode(res.body) as Map<String, dynamic>);
  }

  static Future<dynamic> post(String url, Map<String, dynamic> body) async {
    final res = await http.post(
      Uri.parse(url),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(body),
    );
    if (res.statusCode != 200) {
      throw ApiCryptoException('Server error (${res.statusCode})');
    }
    return ApiCrypto.decryptEnvelope(json.decode(res.body) as Map<String, dynamic>);
  }

  static Future<dynamic> delete(String url) async {
    final res = await http.delete(Uri.parse(url));
    if (res.statusCode != 200) {
      throw ApiCryptoException('Server error (${res.statusCode})');
    }
    return ApiCrypto.decryptEnvelope(json.decode(res.body) as Map<String, dynamic>);
  }
}
